src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,484 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relax/transform/gradient.cc
+ * \brief Reverse-mode automatic differentiation.
+ *
+ * Now only supports differentiating one function in the IRModule with one dataflow block
+ * with respect to the only return value of the function, which needs to be scalar.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/nested_msg.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include <unordered_set>
+
+#include "../op/tensor/binary.h"
+#include "../op/tensor/create.h"
+
+namespace tvm {
+namespace relax {
+
+using AdjointMsg = NestedMsg<Expr>;
+
+// A tool class for GradientMutator
+// Visit the forward bindings and generate the backward bindings
+class BackwardBindingGenerator : public ExprVisitor {
+ public:
+  /*!
+   * \brief Generate the backward bindings for the corresponding GradientMutator
+   *
+   * \param forward_block The forward dataflow block
+   * \param require_grads The relax variables whose adjoints are needed.
+   * \param post The expression with rewritten inputs.
+   */
+  Expr Generate(const BlockBuilder& builder, const DataflowBlock& forward_block,
+                const Array<Var>& require_grads, const Var& target_var) {
+    this->builder_ = builder;
+
+    // initialize the adjoint of target_var as ones op
+    adjoint_msg_map_.Set(target_var, BuildOnesAdjoint(target_var));
+
+    // We do reverse-mode ad, so visit bindings backwards
+    for (auto it = forward_block->bindings.rbegin(); it != forward_block->bindings.rend(); ++it) {
+      this->VisitBinding(*it);
+    }
+
+    return this->Epilogue(require_grads, target_var);
+  }
+
+  void VisitBinding(const Binding& binding) override {
+    // TODO(chaofan, yixin): support other types of bindings
+    CHECK(binding->IsInstance<VarBindingNode>()) << "now only support VarBindingNode";
+    auto var_binding = binding.as<VarBindingNode>();
+
+    if (adjoint_msg_map_.count(var_binding->var) == 0) {
+      // This var is not used in the bindings handled earlier
+      return;
+    }
+
+    // Meet the definition of binding->var
+    // Create the adjoint var and bind the adjoint value to it
+    EmitAdjoint(var_binding->var, adjoint_msg_map_[var_binding->var], true);
```

Comment:
use
```c++
auto it = adjoint_msg_map.find(var_binding->var)
if (it == adjoint_msg_map.end()) {
    return;
}
EmitAdjoint(var_binding->var, (*it).second, true);
```

to avoid duplicate map searching